### PR TITLE
Add a 'list' option for the CLI

### PR DESCRIPTION
**Pull request by abusse**
This commit addresses Issue #448 by adding a command line option that allows the listing of all the Apps reported by the remote host as a CSV.

I wasn't sure what the best output format would be and what information to include. So for the first iteration I went will all available information about an App and the path to the cached box art as CSV. Since it is a simple printf that could be easily adapted if desired.

### CODE REVIEW

File: app/cli/listapps.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+                QCoreApplication::exit(0);
+            }
+            break;
+        }
+    }
+
+    void printApp(NvApp app) const
+    {
+        fprintf(stdout, "%s,%d,%s,%s,%s,%s,%s\n", qPrintable(app.name),
```
Comment (cgutman) [inline app/cli/listapps.cpp:117]: This probably needs to handle escaping the app name, right?

File: app/cli/commandlineparser.cpp [hunk 2]
```diff
@@ -464,3 +467,46 @@ QString StreamCommandLineParser::getAppName() const
...
+{
+    CommandLineParser parser;
+    parser.setupCommonOptions();
+    parser.setApplicationDescription(
+        "\n"
+        "List the available apps on the given host as CSV:\n"
+        "\tName, ID, HDR Support, App Collection Game, Hidden, Direct Launch, Path to Boxart"
+    );
+    parser.addPositionalArgument("list", "list available apps");
+    parser.addPositionalArgument("host", "Host computer name, UUID, or IP address", "<host>");
```
Comment (cgutman) [inline app/cli/commandlineparser.cpp:489]: CSV is nice for consuming in other applications but isn't very friendly for human consumption. I think `list` should just print the app names in a list by default and have a separate `--csv` flag for CSV output.

File: app/gui/CliListAppsSegue.qml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import ComputerManager 1.0
+import Session 1.0
+
+Item {
+    function onSearchingComputer() {
+        stageLabel.text = qsTr("Establishing connection to PC...")
+    }
+
+    function onSearchingApps() {
+        stageLabel.text = qsTr("Searching for Apps...")
```
Comment (cgutman) [inline app/gui/CliListAppsSegue.qml:13]: Let's use `qsTr("Loading app list...")` here like CliStartStreamSegue does.

**Review by cgutman — CHANGES_REQUESTED** [review]

File: app/cli/listapps.cpp (hunk 1 above)
Comment (abusse) [reply app/cli/listapps.cpp:117]: Currently, I cannot think about a situation that is not covered. Could you give an example of an input (app name) that you are concerned about to be escaped?

File: app/gui/CliListAppsSegue.qml (hunk 3 above)
Comment (abusse) [reply app/gui/CliListAppsSegue.qml:13]: ACK

File: app/cli/listapps.cpp (hunk 1 above)
Comment (cgutman) [reply app/cli/listapps.cpp:117]: It's possible for app names to contain commas.

File: app/cli/listapps.cpp [hunk 4]
```diff
@@ -102,25 +102,36 @@ class LauncherPrivate
...
+    void printAppCSV(NvApp app) const
     {
-        fprintf(stdout, "%s,%d,%s,%s,%s,%s,%s\n", qPrintable(app.name),
-                                               app.id,
-                                               app.hdrSupported ? "true" : "false",
-                                               app.isAppCollectorGame ? "true" : "false",
-                                               app.hidden ? "true" : "false",
-                                               app.directLaunch ? "true" : "false",
-                                               qPrintable(m_BoxArtManager->loadBoxArt(m_Computer, app).toDisplayString()));
+        fprintf(stdout, "\"%s\",%d,%s,%s,%s,%s,\"%s\"\n", qPrintable(app.name),
```
Comment (abusse) [inline app/cli/listapps.cpp:128]: One corner case remains if the App name contains a `"` character. I already had code in place to escape that. However, when trying to test, I was not able to have an App with `"` in its name. Therefore, I assume that `"` is not possible in an App name. Furthermore, because I was not able to test the code, I decided to leave it out

**Comment by cgutman** [comment]
I gave this a shot and there are some problems (not really the fault of your change, just how Moonlight currently works) that prevent it from being a good user experience right now IMHO.

1. It doesn't work at all on Windows. We are a Win32 subsystem app which means we don't have a console to write to unless we `AttachConsole()`. I'll need to look into that.
2. Even on platforms where it does work, like macOS and Linux, the output is full of log spam from polling PCs, initializing FFmpeg, etc. I'll need to add some way to maybe increase the minimum log level when running a command like `list` where terminal output is expected.
3. Because a Moonlight window appears when the `list` command is running, this pretty much precludes any background usage by launchers like Playnite. It's also just generally intrusive for human use too. I'm running `moonlight list` from the terminal to write output to that same terminal. Having windows flying around in the process of printing the console is really dumb (again, not your fault, just not something we've plumbed yet).

I'm going to skip this for v3.1.0 due to those issues, but I would like to get these addressed for the following release.

**Comment by abusse** [comment]
I currently don't have a Windows system for testing to my disposal (nor have I done a lot Win32 programming in the past, tbh), so I cannot say much regarding that point. Regarding the second point, I only had this issues with a Debug build. With a Release build the output was "clean". Regarding the third point, that was I concern that I also had, however, since the `quit` command also opened a window, I assumed that was the way to go. I was concerned that a patch "circumventing" this behavior would be too intrusive and not be accepted for a merge. But since we seem to agree on that, I will have another look regarding that direction. Do you have already something in mind? If not, I would just start digging a little further when I have some more time to spare to come to a solution.

**Comment by cgutman** [comment]
Ah yes true, on macOS and Windows release builds, we will redirect the normal error logging to a log file so it won't be printed to the console. I was testing on Linux which prints the output to the terminal.

We can probably use the same state machine and stuff, just printing progress messages to the console instead of using a window.

**Comment by abusse** [comment]
I have made the changes, but I am not sure if that is what you had in mind. Let me know what you think.

**Comment by cgutman** [comment]
Thanks, I'll take a look once I'm back from vacation

**Comment by cgutman** [comment]
Rebased and merged manually to resolve conflicts:

de881769950bf34ed8af6ba8786dcfe378452f6d
0802609fd88341f37c34d197b0a6d569881a18cf
4ee36fd405f382315a494ac8464fb59aa5c44403
